include/knowhere/prometheus_client.h

```diff
@@ -55,38 +55,53 @@ class PrometheusClient {
-DECLARE_PROMETHEUS_HISTOGRAM(knowhere_search_latency);
-DECLARE_PROMETHEUS_HISTOGRAM(knowhere_range_search_latency);
-DECLARE_PROMETHEUS_HISTOGRAM(knowhere_ann_iterator_init_latency);
-DECLARE_PROMETHEUS_HISTOGRAM(knowhere_search_topk);
-
-DECLARE_PROMETHEUS_HISTOGRAM(knowhere_hnsw_bitset_ratio);
-DECLARE_PROMETHEUS_HISTOGRAM(knowhere_hnsw_search_hops);
-
-DECLARE_PROMETHEUS_HISTOGRAM(knowhere_diskann_bitset_ratio);
-DECLARE_PROMETHEUS_HISTOGRAM(knowhere_diskann_search_hops);
-DECLARE_PROMETHEUS_HISTOGRAM(knowhere_diskann_range_search_iters);
+#define CONCATENATE(x, y) x##_##y
+#define PROMETHEUS_LABEL_KNOWHERE knowhere
+#define PROMETHEUS_LABEL_CARDINAL cardinal
+
+#define DEFINE_PROMETHEUS_GAUGE_FAMILY(name, desc)                     \
+    prometheus::Family<prometheus::Gauge>& CONCATENATE(name, family) = \
+        prometheus::BuildGauge().Name(#name).Help(desc).Register(knowhere::prometheusClient->GetRegistry());
+
+#define DEFINE_PROMETHEUS_GAUGE(name, module) \
+    prometheus::Gauge& CONCATENATE(module, name) = CONCATENATE(name, family).Add({{"module", #module}});
+
+#define DEFINE_PROMETHEUS_COUNTER_FAMILY(name, desc)                     \
+    prometheus::Family<prometheus::Counter>& CONCATENATE(name, family) = \
+        prometheus::BuildCounter().Name(#name).Help(desc).Register(knowhere::prometheusClient->GetRegistry());
+
+#define DEFINE_PROMETHEUS_COUNTER(name, module) \
+    prometheus::Counter& CONCATENATE(module, name) = CONCATENATE(name, family).Add({{"module", #module}});
+
+#define DEFINE_PROMETHEUS_HISTOGRAM_FAMILY(name, desc)                     \
+    prometheus::Family<prometheus::Histogram>& CONCATENATE(name, family) = \
+        prometheus::BuildHistogram().Name(#name).Help(desc).Register(knowhere::prometheusClient->GetRegistry());
+
+#define DEFINE_PROMETHEUS_HISTOGRAM_WITH_BUCKETS(name, module, buckets) \
+    prometheus::Histogram& CONCATENATE(module, name) = CONCATENATE(name, family).Add({{"module", #module}}, buckets);
+
+#define DEFINE_PROMETHEUS_HISTOGRAM(name, module) DEFINE_PROMETHEUS_HISTOGRAM_WITH_BUCKETS(name, module, defaultBuckets)
+
+#define DECLARE_PROMETHEUS_GAUGE(name, module) extern prometheus::Gauge& CONCATENATE(module, name);
+#define DECLARE_PROMETHEUS_COUNTER(name, module) extern prometheus::Counter& CONCATENATE(module, name);
+#define DECLARE_PROMETHEUS_HISTOGRAM(name, module) extern prometheus::Histogram& CONCATENATE(module, name);
+
+DECLARE_PROMETHEUS_HISTOGRAM(build_latency, PROMETHEUS_LABEL_KNOWHERE);
+DECLARE_PROMETHEUS_HISTOGRAM(load_latency, PROMETHEUS_LABEL_KNOWHERE);
+
+DECLARE_PROMETHEUS_HISTOGRAM(search_latency, PROMETHEUS_LABEL_KNOWHERE);
+DECLARE_PROMETHEUS_HISTOGRAM(search_latency, PROMETHEUS_LABEL_CARDINAL);
+
+DECLARE_PROMETHEUS_HISTOGRAM(range_search_latency, PROMETHEUS_LABEL_KNOWHERE);
+DECLARE_PROMETHEUS_HISTOGRAM(range_search_latency, PROMETHEUS_LABEL_CARDINAL);
+
+DECLARE_PROMETHEUS_HISTOGRAM(ann_iterator_init_latency, PROMETHEUS_LABEL_KNOWHERE);
```

Comment:
pls help add cardinal label for these ann_iterator_init_latency and search_topk metric

Reply:
done